### ISSUE TITLE
Status Notifier doesn't show fcitx5 icons for languages

### ISSUE THREAD

**Issue by TheHighGroundWins** [qtile/qtile#5122]
### Issue description

Normally on Systray fcitx5 would show the icons of any languages/input methods that use icons instead of text.
Such as MOZC for japanese.
However, on StatusNotifier the icon isn't rendered at all.

Fcitx5 on text based language icons
![2024-12-15-141514_3840x1080_scrot](https://github.com/user-attachments/assets/89346bd5-998f-415d-b294-79804102b770)
Fcitx5 on icon based language icons
![20241215_14h12m58s_grim](https://github.com/user-attachments/assets/5e6a21aa-df08-48f2-a7a7-4f19b636f095)
On Systray:
![2024-12-15-141427_3840x1080_scrot](https://github.com/user-attachments/assets/2a4e78c7-280b-46dc-bc14-de8182b2a2c6)

Maybe the icons have some kind of X11 path. All other icons behave normally

### Version

0.29.1.dev0+geed1e03c.d20241021

### Backend

X11 (default), Wayland (experimental)

### Required

- [X] I have searched past issues to see if this bug has already been reported, and it hasn't been.
- [X] I understand that people give their precious time for free, and thus I've done my very best to make this problem as easy as possible to investigate.

**Comment by elParaguayo** [comment]
This is probably a missing icon issue rather than a bug with the widget.

e.g. https://github.com/qtile/qtile/issues/4503#issuecomment-1752042396

**Comment by TheHighGroundWins** [comment]
Seems 

> This is probably a missing icon issue rather than a bug with the widget.
> 
> e.g. [#4503 (comment)](https://github.com/qtile/qtile/issues/4503#issuecomment-1752042396)

getIconPath('input-keyboard') returns nothing.
using `StatusNotifier(icon_theme="Breeze"),`

**Comment by TheHighGroundWins** [comment]
When I restart fcitx5, the opposite happens. The icon based icons display, but the text based icons don't
![20241215_21h46m05s_grim](https://github.com/user-attachments/assets/18e622f1-2c0a-461b-a662-7a79ed646d60)
![20241215_21h44m53s_grim](https://github.com/user-attachments/assets/1e1c7287-3cb0-4cd7-b1f8-aad48bdabd98)

**Comment by elParaguayo** [comment]
Does the log say which icons it can't find?

**Comment by TheHighGroundWins** [comment]
`2024-12-15 21:57:47,145 WARNING libqtile statusnotifier.py:_get_local_icon():L264 Could not find icon for ''. Using fallback icon.`
`2024-12-15 22:00:23,347 WARNING libqtile statusnotifier.py:_get_local_icon():L264 Could not find icon for ''. Using fallback icon.`
This message repeated over and over again.

It shows the text based icons on first boot or when closed and opened. But the other icon based icons was the widget is restarted within fcitx5 itself.

**Comment by elParaguayo** [comment]
Oh, that doesn't look good. There doesn't seem to be an icon name.

**Comment by TheHighGroundWins** [comment]
> Oh, that doesn't look good. There doesn't seem to be an icon name.

This seems to happens only for the letter based icons. So there is probably no icon in there just letters.

**Comment by elParaguayo** [comment]
Sure, but the app still needs to indicate what it wants us to display. There's nothing in the spec for plain text.

**Comment by TheHighGroundWins** [comment]
```
2024-12-17 16:44:56,582 ERROR libqtile loop.py:_handle_exception():L62 Exception in event loop:`
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/libqtile/widget/helpers/status_notifier/statusnotifier.py", line 336, in _get_icon
    icon_pixmap = await pixmap()
                  ^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/dbus_next/aio/proxy_object.py", line 118, in property_getter
    BaseProxyInterface._check_method_return(msg, 'v')
  File "/usr/lib/python3.12/site-packages/dbus_next/proxy_object.py", line 62, in _check_method_return
    raise DBusError._from_message(msg)
```
I found this other error message.

I realized I didn't have pyxdg installed. Now its a little better, the icons are displaying more consistently now with the non english icons displaying normally, but still no icons for the letters.

**Comment by TheHighGroundWins** [comment]
Oh I excluded the dbus errors, didn't think that they were important. Yeah I get not wanting to add extra languages, I've had situation where I can't navigate qtile due to my current language and was confused for a second.

Heres the error message with the DBus errors:
```
2024-12-17 16:44:56,582 ERROR libqtile loop.py:_handle_exception():L62 Exception in event loop:
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/libqtile/widget/helpers/status_notifier/statusnotifier.py", line 336, in _get_icon
    icon_pixmap = await pixmap()
                  ^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/dbus_next/aio/proxy_object.py", line 118, in property_getter
    BaseProxyInterface._check_method_return(msg, 'v')
  File "/usr/lib/python3.12/site-packages/dbus_next/proxy_object.py", line 62, in _check_method_return
    raise DBusError._from_message(msg)
dbus_next.errors.DBusError: No such object path '/StatusNotifierItem'
```

**Comment by elParaguayo** [comment]
Well that's an odd error which shouldn't happen.

Is it possible for me to use `fcitx` but still type in English? I think I need to try to recreate this.

**Comment by TheHighGroundWins** [comment]
> Well that's an odd error which shouldn't happen.
> 
> Is it possible for me to use `fcitx` but still type in English? I think I need to try to recreate this.

That's strange. Oh yes definitely unless you explicitly install another language fcitx5 should just have English or whatever language you use by default.



Note:
On startup it should show a keyboard icon, so if there's an empty space that the missing icon.

If the default logo for English does come up try right clicking the tray icon and pressing restore. That triggers the bug where the other type of icon isn't showing up.

It only shows one of the other type of icon.

**Comment by elParaguayo** [comment]
I'm struggling with this but probably because I've got no idea how to use fcitx5.

Could you try something for me:
- Close all other apps that create an icon in StatusNotifier
- Run `busctl get-property --user org.kde.StatusNotifierWatcher /StatusNotifierWatcher org.kde.StatusNotifierWatcher RegisteredStatusNotifierItems` to get the address of fcitx5 on the bus. Example output for me was `as 1 :1.44`
- Get fcitx5 in a position where the text icon should be displayed
- Run `busctl introspect --user :1.44 /StatusNotifierItem` (changing `:1.44` for whatever you got in the above command)

**Comment by TheHighGroundWins** [comment]
Thank you for sticking through this process btw.

I did everything as you said and also put the busctl for when the icon is and isn't displaying
```
as 1 ":1.54"
```

When icon is shown:
```
NAME                                TYPE      SIGNATURE    RESULT/VALUE     FLAGS
org.freedesktop.DBus.Introspectable interface -            -                -
.Introspect                         method    -            s                -
org.freedesktop.DBus.Properties     interface -            -                -
.Get                                method    ss           v                -
.GetAll                             method    s            a{sv}            -
.Set                                method    ssv          -                -
.PropertiesChanged                  signal    sa{sv}as     -                -
org.kde.StatusNotifierItem          interface -            -                -
.Activate                           method    ii           -                -
.Scroll                             method    is           -                -
.SecondaryActivate                  method    ii           -                -
.AttentionIconName                  property  s            ""               emits-change
.AttentionIconPixmap                property  a(iiay)      0                emits-change
.AttentionMovieName                 property  s            ""               emits-change
.Category                           property  s            "SystemServices" emits-change
.IconAccessibleDesc                 property  s            "Input Method"   emits-change
.IconName                           property  s            "mn_trad"        emits-change
.IconPixmap                         property  a(iiay)      0                emits-change
.IconThemePath                      property  s            ""               emits-change
.Id                                 property  s            "Fcitx"          emits-change
.ItemIsMenu                         property  b            false            emits-change
.Menu                               property  o            "/MenuBar"       emits-change
.OverlayIconName                    property  s            ""               emits-change
.OverlayIconPixmap                  property  a(iiay)      1 1 1 4 0 0 0 0  emits-change
.Status                             property  s            "Active"         emits-change
.Title                              property  s            "Input Method"   emits-change
.ToolTip                            property  (sa(iiay)ss) "" 0 "" ""       emits-change
.WindowId                           property  i            0                emits-change
.XAyatanaLabel                      property  s            ""               emits-change
.XAyatanaLabelGuide                 property  s            ""               emits-change
.XAyatanaOrderingIndex              property  u            0                emits-change
.NewAttentionIcon                   signal    -            -                -
.NewIcon                            signal    -            -                -
.NewIconThemePath                   signal    s            -                -
.NewOverlayIcon                     signal    -            -                -
.NewStatus                          signal    s            -                -
.NewTitle                           signal    -            -                -
.NewToolTip                         signal    -            -                -
.XAyatanaNewLabel                   signal    ss           -                -
```
When Icon isn't showing:
```
NAME                                TYPE      SIGNATURE    RESULT/VALUE                             FLAGS
org.freedesktop.DBus.Introspectable interface -            -                                        -
.Introspect                         method    -            s                                        -
org.freedesktop.DBus.Properties     interface -            -                                        -
.Get                                method    ss           v                                        -
.GetAll                             method    s            a{sv}                                    -
.Set                                method    ssv          -                                        -
.PropertiesChanged                  signal    sa{sv}as     -                                        -
org.kde.StatusNotifierItem          interface -            -                                        -
.Activate                           method    ii           -                                        -
.Scroll                             method    is           -                                        -
.SecondaryActivate                  method    ii           -                                        -
.AttentionIconName                  property  s            ""                                       emits-change
.AttentionIconPixmap                property  a(iiay)      0                                        emits-change
.AttentionMovieName                 property  s            ""                                       emits-change
.Category                           property  s            "SystemServices"                         emits-change
.IconAccessibleDesc                 property  s            "Input Method"                           emits-change
.IconName                           property  s            ""                                       emits-change
.IconPixmap                         property  a(iiay)      4 16 16 1024 0 0 0 0 0 0 0 0 0 0 0 0 0 … emits-change
.IconThemePath                      property  s            ""                                       emits-change
.Id                                 property  s            "Fcitx"                                  emits-change
.ItemIsMenu                         property  b            false                                    emits-change
.Menu                               property  o            "/MenuBar"                               emits-change
.OverlayIconName                    property  s            ""                                       emits-change
.OverlayIconPixmap                  property  a(iiay)      1 1 1 4 0 0 0 0                          emits-change
.Status                             property  s            "Active"                                 emits-change
.Title                              property  s            "Input Method"                           emits-change
.ToolTip                            property  (sa(iiay)ss) "" 0 "" ""                               emits-change
.WindowId                           property  i            0                                        emits-change
.XAyatanaLabel                      property  s            ""                                       emits-change
.XAyatanaLabelGuide                 property  s            ""                                       emits-change
.XAyatanaOrderingIndex              property  u            0                                        emits-change
.NewAttentionIcon                   signal    -            -                                        -
.NewIcon                            signal    -            -                                        -
.NewIconThemePath                   signal    s            -                                        -
.NewOverlayIcon                     signal    -            -                                        -
.NewStatus                          signal    s            -                                        -
.NewTitle                           signal    -            -                                        -
.NewToolTip                         signal    -            -                                        -
.XAyatanaNewLabel                   signal    ss           -                                        -


```

**Comment by TheHighGroundWins** [comment]
I tried testing with only english and seems like the bug only occurs when there is more than one language available.

So I got the missing icon bug to happen by having both US english and UK english keyboards. They show up as separate languages.

You can install fcitx5-configtool to use the GUI to and remove languages, which is how I personally do it.

If the bug still isn't happening try either try clicking on the fcitx5 widget and pressing Exit, and reopening fcitx5 by just starting fcitx5 in the terminal, dmenu or whichever program you prefer. This seems to trigger the switch to DBUS for me.
To switch to local icons, right click on the fcitx5 widget and press restore.

**Comment by elParaguayo** [comment]
OK. So I recreated this, briefly, once. However, usually when I look at the menu, neither of the languages is selected and clicking on one has no impact. The one time I could recreate the issue, the menu showed one language as selected.

**Comment by TheHighGroundWins** [comment]
Maybe because the environment variables for fcitx5 haven't been set?
You need these in your environment variables:
```
XMODIFIERS '@im=fcitx'
GTK_IM_MODULE 'fcitx'
QT_IM_MODULE 'fcitx5'
```
Should work on both wayland and X11, except on wayland terminals.

If the icons are still showing correctly try pressing restart on the fcitx5 tray. Or press Exit and reopen. The behavior is quite random on my end as well maybe race condition.
![20241220_16h55m18s_grim](https://github.com/user-attachments/assets/7fc8fb10-bb8d-48c9-bd58-c37a577fa677)

**Comment by elParaguayo** [comment]
OK - I know what the issue is.

Basically, if the widget uses a local icon, when it receives a `NewIcon` signal, it tries to update the local icon without checking to see if the icon is now being provided by the app itself.

I can work on a fix for this.

**Comment by elParaguayo** [comment]
Can you try the fix in #5128 please.

**Comment by TheHighGroundWins** [comment]
> Can you try the fix in #5128 please.

Just finished compiling and running it. Now all the different icons work without an issue thanks.

**Comment by elParaguayo** [comment]
Great. Thanks for checking. I've merged this into the master branch.
